_posts/2021-06-28-regular-expressions-in-ruby.md

```diff
@@ -0,0 +1,146 @@
+
+And now we will introduce a restriction on the last letter, that is, the letter `t` must be the last in the line, otherwise the line will not pass our 'funnel'. To do this, we will use an anchor `$`- denoting the end of the line:
+
+```ruby
+'cat' =~ /^c.t$/
+=> 0
+
+'cute' =~ /^c.t$/
+=> nil
+```
+
+In the last example, the letter is tnot the last character, hence the result nil.
+
+It's time to get acquainted with the quantifiers ( Quantifier ) from the English word Quantify - which means to determine the amount . Let's take an example already familiar to us, and substitute an asterisk symbol *( asterisk , asterisk ) instead of a dot . Using this quantifier, we can enter any number of characters from zero to infinity in place of the asterisk:
+
+```ruby
+'catttttttttttttttt' =~ /^c.t*$/
+=> 0
+
+'ca' =~ /^c.t*$/
+=> 0
+```
+
+Let's validate the email, that is, create a simplified regular expression that checks the correctness of the entered email:
+
+```ruby
+/^[a-z0-9]+@[a-z0-9]+\.[a-z]+/ =~ 'google@gmail.com'
+=> 0
+
+/^[a-z0-9]+@[a-z0-9]+\.[a-z]+/ =~ 'google@gmail.%%'
+=> nil
+```
+
+Let's take a closer look at the example:
+
+* `//` - as we remember, we put our regular expression between the slash characters
+
+* `^` - this anchor means the beginning of a line
+
+* `[a-z0-9]`- character class, here we indicated that any numeric character is allowed, as well as any lowercase (small, not uppercase) letter of the English alphabet. Accordingly, if we wanted to skip capital letters, then we would write `[A-Z]`. Or, to make it even more beautiful and more compact at the end of our regular expression we can add a modifier ( modifier ) `i`, which turns a blind eye on a case letters (skip both uppercase and lowercase). The expression will take the form:`/^[a-z0-9]+@[a-z0-9]+\.[a-z]+/i`
+
+* `+` - a quantifier that skips a result from one to infinity, that is, in our case `[a-z0-9]+`, any number of letters of the English alphabet and numbers from 1 to infinity are allowed
+
+* `@` - check that the string has the given character in the strictly specified place, and this character must be no more and no less than one
+
+* `\.` - we remember that .in a regular expression it means that in place of a given dot we can use any character in the singular, but we want to specify a specific dot character that is written in the email, for this we need to escape our dot with a backslash
+
+`[a-z]+` - and lastly, we check the spelling of the top-level domain name (that is, for example, `ru` or `com`)
+
+We can write our regular expression a little differently using the following syntax:
+
+```ruby
+`/^[\w\d]+@[\w\d]+\.[\w]+/ =~ 'google2012@gmail.com'`
+=> 0
+```
+
+Where
+
+`\w`- similar to an expression `[a-zA-Z]`, that is, any letter of the English alphabet, independent of the case, but in addition to this, the presence of an underscore is also allowed `_`, which in general is what we need to check email, since this character in an email address is acceptable
```

Comment:
```suggestion
`\w`- similar to an expression `[a-zA-Z]`, that is, any letter of the English alphabet, independent of the case, but in addition to this, the presence of an underscore is also allowed `_`, which in general is what we need to check email since this character in an email address is acceptable.
```